### PR TITLE
PyGRB Workflow Generation Updates

### PR DESCRIPTION
**Pull request by jakeb245**
This contains the changes made needed to make the PyGRB workflow generator work with `pycbc_multi_inspiral`. Currently, a simple workflow containing splitbank and inspiral jobs is able to be generated and successfully ran with the right configuration options. 

Some of the code was already a bit rough, and I haven't done a lot to fix that yet. I may have even made it worse at times, but it does work. Therefore I am marking this PR as work in progress as I'd like to clean some things up before anything is merged. Advice on how to do that would be appreciated! I tried to make it work with the original coh_ptf code for now, but dropping that would make things simpler. 

There are also some fixes to `pycbc_multi_inspiral` within this PR. One of them was a fault of myself not noticing a check in #3798. My fix to that here works (I think), but I intend to come up with a better way of doing it. I also changed the call to the auto chi-squared test so that it uses the SNR timeseries like it's supposed to. Before, it was using the list of trigger SNRs and sometimes causing an index error. I'm considering moving these fixes to a separate pull request.

### CODE REVIEW

File: pycbc/workflow/matched_filter.py
Comment (titodalcanton) [inline pycbc/workflow/matched_filter.py:243]:
```suggestion
        cp.set('inspiral', 'ra',
```

File: pycbc/workflow/matched_filter.py
Comment (titodalcanton) [inline pycbc/workflow/matched_filter.py:245]:
```suggestion
        cp.set('inspiral', 'dec',
```

File: pycbc/workflow/segment.py
Comment (titodalcanton) [inline pycbc/workflow/segment.py:312]:
```suggestion

```

File: pycbc/workflow/jobsetup.py
Comment (titodalcanton) [inline pycbc/workflow/jobsetup.py:711]:
```suggestion

```

**Comment by jakeb245** [comment]
Since creating this PR I've made some progress. Currently the code is able to generate a PyGRB workflow running the following jobs in order.
1. Splitbank
2. Inspiral (no injections)
3. Trig combiner
4. Trig cluster

The rest of the PyGRB workflow is still under development by others. I think this PR is in a state where it can be merged (after review) with more changes coming in new PRs as other parts of the workflow are finished.

**Comment by jakeb245** [comment]
I'm going to be unable to work on this for several weeks, so I'll give an update on this PR and then respond to reviews until it can be merged.

This PR brings the PyGRB workflow generation code to a state where it can run the new executables until post-processing plots/efficiency. See [this wiki](https://github.com/gwastro/pycbc/wiki/PyGRB) for more information. Some general issues I've found are:
- `pycbc_grb_inj_combiner` does not work with the output of `pycbc_grb_inj_finder`. The file structures do not seem to be compatible as written. The workflow generates and runs both executables without failure, but the combiner does not do anything.
- Currently, the config files given to `pycbc_create_injections` must be hardcoded for the event-specific parameters. In the future, we likely want the workflow generator to dynamically generate config files with those settings. Getting EM bright and jitter skyloc functionality merged into master should likely happen before that.
- CodeClimate is failing, but the issues don't seem worth fixing to me. I don't quite understand the issues, so I'm happy to fix them if requested.

Some of these issues are outside the scope of the PyGRB workflow generator and can be fixed in seperate PRs. As more updates to the PyGRB workflow are finalized, more updates to the workflow generator will be required. These updates should cover all work that has been finished and merged into master. There are also some small fixes in the executables required for them to run.

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:11]: Move the docstring between the shebang and the imports.

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:20]:
```suggestion
parser.add_argument("-f", "--output-files", nargs='*', required=True,
                    help="Names of output files")
parser.add_argument("-i", "--input-file", required=True,
                    help="Injection file to be split")
```

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:25]:
```suggestion
inj_file = h5py.File(args.input_file, 'r')
```

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:39]: You can probably drop the `.keys()` here.

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:57]: An assert is probably better than a ValueError here:
```suggestion
assert sum(injs_per_split) == num_injs, "Not all injections were accounted for!"
```

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:48]:
```suggestion
injs_per_split = np.zeros(num_splits, dtype=int) + ideal_split
```

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [inline bin/pycbc_hdf_splitinj:65]:
```suggestion
    injs_in_split = injs_per_split[i]
```

File: bin/pycbc_hdf_splitinj
Comment (titodalcanton) [reply bin/pycbc_hdf_splitinj:11]: After seeing the comment towards the end of the script, I would also add a sentence to the docstring explaining that the split injections are organized so as to maximize the time between injections in the output files. Otherwise, the user may simply assume that the first output file will get the first N injections in the input file, etc.

File: bin/pycbc_multi_inspiral
Comment (titodalcanton) [inline bin/pycbc_multi_inspiral:353]: Since I am not familiar with the code around this change, what is the reason for the change? Was the old logic incorrect? Should the code also check that `idx[ifo] + time_delay_idx[ifo] >= 0`?

File: bin/pygrb/pycbc_grb_trig_combiner
Comment (titodalcanton) [inline bin/pygrb/pycbc_grb_trig_combiner:483]:
```suggestion
    _ts += j * trialtime
```

File: bin/pygrb/pycbc_make_offline_grb_workflow
Comment (titodalcanton) [inline bin/pygrb/pycbc_make_offline_grb_workflow:81]:
```suggestion
sciSegsFile = _workflow.get_segments_file(wflow, 'science', 'segments-science', segDir)
```

File: pycbc/workflow/grb_utils.py
Comment (titodalcanton) [inline pycbc/workflow/grb_utils.py:75]:
```suggestion
    if exe_name not in exe_to_class_map:
        raise ValueError(f"No job class exists for executable {curr_exe}")
```

File: pycbc/workflow/grb_utils.py
Comment (titodalcanton) [inline pycbc/workflow/grb_utils.py:438]: Is the `[1]` index intended on the last line (and not intended a couple of lines before)? If so, consider adding a comment explaining why you only look at the second element.

File: pycbc/workflow/grb_utils.py
Comment (titodalcanton) [inline pycbc/workflow/grb_utils.py:488]: F-strings are great, use them:
```suggestion
        user_tag = f"PYGRB_GRB{self.trigger_name}"
```

File: pycbc/workflow/grb_utils.py
Comment (titodalcanton) [inline pycbc/workflow/grb_utils.py:575]: This sounds like something that we absolutely want in a config file and not hardcoded here, maybe add a FIXME comment.

File: pycbc/workflow/injection.py
Comment (titodalcanton) [inline pycbc/workflow/injection.py:230]: Does this actually work, or did you intend to use `isinstance()`?

File: pycbc/workflow/jobsetup.py
Comment (titodalcanton) [inline pycbc/workflow/jobsetup.py:298]: This function seems particularly fragile (and the comment just above the change appears to agree). Looping over `datafind_outs` until you find one with the right name seems a better way, would that not work?

File: pycbc/workflow/jobsetup.py
Comment (titodalcanton) [inline pycbc/workflow/jobsetup.py:721]: Much simpler with an f-string:
```suggestion
                frame_arg += f' {frame_file.ifo}:{frame_file.name}'
```

File: pycbc/workflow/matched_filter.py
Comment (titodalcanton) [inline pycbc/workflow/matched_filter.py:245]: Does this change imply that the config files will now have to specify RA and declination in radians, or that the inspiral executable now takes degrees?

File: bin/pycbc_multi_inspiral
Comment (jakeb245) [reply bin/pycbc_multi_inspiral:353]: The time delay was previously applied using a numpy operation. It was changed so that time delay is now added as necessary. The logic check here was overlooked when that change was made, so when I was running workflow with many triggers it would try and run chi-squared tests on triggers with indices outside of the data. There certainly could be cases where more logic checks are necessary.

File: bin/pygrb/pycbc_grb_inj_finder
Comment (jakeb245) [reply bin/pygrb/pycbc_grb_inj_finder:322]: I had to do this because h5py does not seem to support Unicode data types. It works, but there may be a better way to work around it.

File: pycbc/workflow/injection.py
Comment (jakeb245) [reply pycbc/workflow/injection.py:230]: I probably did intend to use `isinstance()` but this does work

File: pycbc/workflow/jobsetup.py
Comment (jakeb245) [reply pycbc/workflow/jobsetup.py:298]: This code is definitely questionable. I can look into rewriting it.

File: pycbc/workflow/matched_filter.py
Comment (jakeb245) [reply pycbc/workflow/matched_filter.py:245]: I believe I changed this because `pycbc_multi_inspiral` was changed to use RA and declination in radians.

File: pycbc/workflow/injection.py
Comment (titodalcanton) [reply pycbc/workflow/injection.py:230]: I am not convinced:
```
In [1]: class Test:
   ...:     def __init__(self):
   ...:         self.foo = 12
   ...: 

In [2]: x = Test()

In [3]: isinstance(x, Test)
Out[3]: True

In [4]: x is Test
Out[4]: False
```

Comment (titodalcanton) [reply pycbc/workflow/injection.py:230]: Ah sorry, `select_generic_executable()` returns the actual class, not an instance. So yeah, this would work.

File: bin/pygrb/pycbc_grb_inj_finder
Comment (titodalcanton) [reply bin/pygrb/pycbc_grb_inj_finder:322]: I get the h5py and Unicode issue. I think my issue with the last line is the following. `injs[field]` would be a Numpy array of strings here. What you want to do is convert the Numpy array of strings to a list of Python strings, which h5py will support. However, the last line will not do that. It will create a *single* string, containing just how Numpy would print the array. This should do what you want:
```suggestion
                grp[field] = [str(s) for s in injs[field]]
```

File: pycbc/workflow/splittable.py
Comment (titodalcanton) [inline pycbc/workflow/splittable.py:154]: This seems to require a syntax change in the configuration file. It would be good if the new workflow could run with old configuration files as much as possible. Given that this is a trivial change, can we avoid it for now?

File: pycbc/workflow/matched_filter.py
Comment (jakeb245) [reply pycbc/workflow/matched_filter.py:245]: I believe I was thinking that with `pycbc_multi_inspiral` now taking RA and dec, it could be directly provided from the config file without any conversion. Looking at it again, I'm not sure if that's correct. I'll have to look into this further.

File: pycbc/workflow/grb_utils.py
Comment (jakeb245) [reply pycbc/workflow/grb_utils.py:437]: I think this was because the first file in the inspiral file list is (but is not always) the bank veto bank. Looking at it now, I can probably come up with a better way to get around it.

Comment (titodalcanton) [reply pycbc/workflow/grb_utils.py:437]: Any update on this?

File: pycbc/workflow/matched_filter.py
Comment (titodalcanton) [reply pycbc/workflow/matched_filter.py:245]: Has this been looked into?

Comment (titodalcanton) [reply pycbc/workflow/matched_filter.py:245]: On the call, we decided to keep the conversion from degrees (easy for whoever sets up the .ini file from a GRB) to radians (which is what `pycbc_multi_inspiral` understands).

File: pycbc/workflow/grb_utils.py
Comment (titodalcanton) [reply pycbc/workflow/grb_utils.py:437]: Please check this or at least add a FIXME comment so we do not forget.

File: pycbc/workflow/splittable.py
Comment (jakeb245) [reply pycbc/workflow/splittable.py:156]: I remembered why I did this... it was because `pycbc_create_injections` uses this variable name. I've reverted it and a comment clarifying this reasoning. I can also add a check for the injections executable, or there could be a better way to handle it.

Comment (titodalcanton) [reply pycbc/workflow/splittable.py:156]: Ok.

**Review by titodalcanton — APPROVED** [review]
Thanks @jakeb245, I think it is now time to get this merged so development can continue.

**Pull request merged**
